service/src/main/java/uk/gov/hmcts/reform/fpl/service/ApplicantMigrationService.java

```diff
@@ -0,0 +1,150 @@
+        if (caseData.getApplicants() == null) {
+            List<Element<Applicant>> populatedApplicant = new ArrayList<>();
+
+            populatedApplicant.add(Element.<Applicant>builder()
+                .value(Applicant.builder()
+                    .party(ApplicantParty.builder()
+                        .partyId(UUID.randomUUID().toString())
+                        .build())
+                    .build())
+                .build());
+
+            return populatedApplicant;
+        } else {
+            return caseData.getApplicants();
+        }
+    }
+
+    public AboutToStartOrSubmitCallbackResponse validateAndUpdatePBANumbers(CaseDetails caseDetails) {
+        ImmutableList.Builder<String> validationErrors = ImmutableList.builder();
+        CaseData caseData = mapper.convertValue(caseDetails.getData(), CaseData.class);
+
+        if (caseData.getApplicants() != null) {
+            List<Element<Applicant>> applicants = caseData.getApplicants().stream()
+                .map(element -> {
+                    Applicant.ApplicantBuilder applicantBuilder = Applicant.builder();
+
+                    if (element.getValue().getParty().getPbaNumber() != null) {
+                        String pba = updatePBANumber(element.getValue().getParty().getPbaNumber());
+                        validationErrors.addAll(validatePBANumber(pba));
+
+                        applicantBuilder.party(element.getValue().getParty().toBuilder().pbaNumber(pba).build());
+                    }
+
+                    return Element.<Applicant>builder()
+                        .id(element.getId())
+                        .value(applicantBuilder.build())
+                        .build();
+                })
+                .collect(Collectors.toList());
+
+            caseDetails.getData().put("applicants", applicants);
+
+        } else {
+            OldApplicant applicantData = caseData.getApplicant();
+
+            if (isNullOrEmpty(applicantData.getPbaNumber())) {
+                return AboutToStartOrSubmitCallbackResponse.builder()
+                    .data(caseDetails.getData())
+                    .errors(validationErrors.build())
+                    .build();
+            }
+
+            String newPbaNumberData = PBANumberHelper.updatePBANumber(applicantData.getPbaNumber());
+            validationErrors.addAll(validatePBANumber(newPbaNumberData));
+
+            if (validationErrors.build().isEmpty()) {
+                applicantData.setPbaNumber(newPbaNumberData);
+                caseDetails.getData().put("applicant", applicantData);
+            }
+        }
+        return AboutToStartOrSubmitCallbackResponse.builder()
+            .data(caseDetails.getData())
+            .errors(validationErrors.build())
+            .build();
+    }
+
+    @SuppressWarnings("unchecked")
+    public AboutToStartOrSubmitCallbackResponse addHiddenValues(CaseDetails caseDetails) {
+        Map<String, Object> data = caseDetails.getData();
```

Comment:
map to CaseData, and remove all the List<Map>. Use our object

Reply:
👍